### PR TITLE
fix asynchronous version of Filter.get().

### PR DESCRIPTION
**Pull request by bethnull**
Calls to get(callback) are postponed while filterId is not available.

### CODE REVIEW

File: lib/web3/filter.js [hunk 1]
```diff
@@ -149,6 +150,13 @@ var Filter = function (options, methods, formatter, callback) {
         } else {
             self.filterId = id;
 
+            // check if there are get pending callbacks as a consequence
+            // of calling get() with filterId unassigned.
+            self.getPendingCallbacks.forEach(function (cb){
```
Comment (frozeman) [inline lib/web3/filter.js]: I would like to rename `getPendingCallbacks` to `getLogsCallbacks`

Apart from that it looks nice!
Did you tested it?

**Comment by frozeman** [comment]
Ideal would also be a test :)
Sorry for the late response, i was on holiday

File: lib/web3/filter.js (hunk 1 above)
Comment (bethnull) [inline lib/web3/filter.js]: Hi, no problem with the renaming ;)

Yes I used it for a few days and worked, however I started to notice that sometimes not all logs were retrieved. Sometimes get() returned all the logs, sometimes one was missing, another times a few and in a pretty erratic pattern. To make things worse, the missing logs were not always the same :(.

The only explanation I can find is that this.implementation.getLogs(this.filterId, function(err, res) is not properly working.

I'll try to upload a test this afternoon/night to check that the asynchronous call of get is working and try to replicate the missing logs behaviour.

Comment (frozeman) [inline lib/web3/filter.js]: Could you make the name change :)

could you check in your network tap, if he missing logs are also missing in the response from `eth_getFilterLogs`?

Maybe they are dropped by the log parser or something @debris any idea?

File: lib/web3/filter.js [hunk 2]
```diff
@@ -187,15 +195,21 @@ Filter.prototype.stopWatching = function () {
...
+                    callback(err);
+                } else {
+                    callback(null, res.map(function (log) {
+                        return self.formatter ? self.formatter(log) : log;
+                    }));
+                }
+            });
+        }
     } else {
         var logs = this.implementation.getLogs(this.filterId);
```
Comment (frozeman) [inline lib/web3/filter.js]: You actually also should throw an error here when no filter id is set, as chaining sync is not possible anymore.
like:
```js
if(this.filterId === null) {
    throw new Error('Filter ID Error: filter().get() can\'t be chained synchronous, please provide a callback for the get() method.');
}
```

Comment (bethnull) [inline lib/web3/filter.js]: Do you mean like this?

```javascript
} else {
        if (this.filterId === null) {
            throw new Error('Filter ID Error: filter().get() can\'t be chained synchronous, please provide a callback for the get() method.');
        }
        var logs = this.implementation.getLogs(this.filterId);
        return logs.map(function (log) {
            return self.formatter ? self.formatter(log) : log;
        });
    }
```
Allowing the synchronous call if the filterId is present?

Comment (frozeman) [inline lib/web3/filter.js]: :+1:

**Comment by frozeman** [comment]
@bethnull  You can add the tests to  this file https://github.com/ethereum/web3.js/blob/develop/test/web3.eth.filter.js

And then do something like:
- test if it throws an error on sync
- use .get and see if it calls the right function

Try to modify the file like this and see if it tests correctly (i didn't test this)
```js
var chai = require('chai');
var web3 = require('../index');
var assert = chai.assert;
var FakeHttpProvider = require('./helpers/FakeHttpProvider');

var method = 'filter';


var tests = [{
    args: [{
        fromBlock: 0,
        toBlock: 10,
        address: '0x47d33b27bb249a2dbab4c0612bf9caf4c1950855'
    }],
    formattedArgs: [{
        fromBlock: '0x0',
        toBlock: '0xa',
        address: '0x47d33b27bb249a2dbab4c0612bf9caf4c1950855',
        topics: []
    }],
    result: '0xf',
    formattedResult: '0xf',
    call: 'eth_newFilter'
},{
    args: [{
        fromBlock: 'latest',
        toBlock: 'latest',
        address: '0x47d33b27bb249a2dbab4c0612bf9caf4c1950855'
    }],
    formattedArgs: [{
        fromBlock: 'latest',
        toBlock: 'latest',
        address: '0x47d33b27bb249a2dbab4c0612bf9caf4c1950855',
        topics: []
    }],
    result: '0xf',
    formattedResult: '0xf',
    call: 'eth_newFilter'
},{
    args: ['latest'],
    formattedArgs: [],
    result: '0xf',
    formattedResult: '0xf',
    call: 'eth_newBlockFilter'
},{
    args: ['pending'],
    formattedArgs: [],
    result: '0xf',
    formattedResult: '0xf',
    call: 'eth_newPendingTransactionFilter'
}];

describe('web3.eth', function () {
    describe(method, function () {
        tests.forEach(function (test, index) {
            it('property test: ' + index, function () {
                
                // given
                var provider = new FakeHttpProvider();
                web3.setProvider(provider);
                provider.injectResult(test.result);
                provider.injectValidation(function (payload) {
                    assert.equal(payload.jsonrpc, '2.0');
                    assert.equal(payload.method, test.call);
                    assert.deepEqual(payload.params, test.formattedArgs);
                });

                // call
                var filter = web3.eth[method].apply(null, test.args);
                

                // test filter.get
                if(typeof test.args === 'object') {
                    
                    var logs = [{data: '0xb'}, {data: '0x11'}];

                    provider.injectResult(logs);
                    provider.injectValidation(function (payload) {
                        assert.equal(payload.jsonrpc, '2.0');
                        assert.equal(payload.method, 'eth_getFilterLogs');
                        assert.deepEqual(payload.params, test.formattedResult);
                    });

                    // should thorw error
                    assert.throws(filter.get());

                    // should get the fake logs
                    filter.get(function(e, res){
                        assert.equal(logs, res);
                    });
                }
            });
        });
    });
});
```

**Comment by bethnull** [comment]
Hi Fabian, sorry for the delay.

I've tried your tests receiving the following errors:

```
592 passing (4s)
  4 failing

  1) web3.eth filter property test: 0:
     AssertionError: expected [ '0xf' ] to deeply equal '0xf'
      at null.validation (test/web3.eth.filter.js:80:32)
      at FakeHttpProvider.send (test/helpers/FakeHttpProvider.js:38:14)
      at RequestManager.send (lib/web3/requestmanager.js:73:32)
      at Method.send (lib/web3/method.js:168:59)
      at Filter.get (lib/web3/filter.js:217:40)
      at Context.<anonymous> (test/web3.eth.filter.js:84:42)

  2) web3.eth filter property test: 1:
     AssertionError: expected [ '0xf' ] to deeply equal '0xf'
      at null.validation (test/web3.eth.filter.js:80:32)
      at FakeHttpProvider.send (test/helpers/FakeHttpProvider.js:38:14)
      at RequestManager.send (lib/web3/requestmanager.js:73:32)
      at Method.send (lib/web3/method.js:168:59)
      at Filter.get (lib/web3/filter.js:217:40)
      at Context.<anonymous> (test/web3.eth.filter.js:84:42)

  3) web3.eth filter property test: 2:
     AssertionError: expected [ '0xf' ] to deeply equal '0xf'
      at null.validation (test/web3.eth.filter.js:80:32)
      at FakeHttpProvider.send (test/helpers/FakeHttpProvider.js:38:14)
      at RequestManager.send (lib/web3/requestmanager.js:73:32)
      at Method.send (lib/web3/method.js:168:59)
      at Filter.get (lib/web3/filter.js:217:40)
      at Context.<anonymous> (test/web3.eth.filter.js:84:42)

  4) web3.eth filter property test: 3:
     AssertionError: expected [ '0xf' ] to deeply equal '0xf'
      at null.validation (test/web3.eth.filter.js:80:32)
      at FakeHttpProvider.send (test/helpers/FakeHttpProvider.js:38:14)
      at RequestManager.send (lib/web3/requestmanager.js:73:32)
      at Method.send (lib/web3/method.js:168:59)
      at Filter.get (lib/web3/filter.js:217:40)
      at Context.<anonymous> (test/web3.eth.filter.js:84:42)
``` 
I'm sorry but I still don't fully understand how the testing works and why assert.deepEqual(payload.params, test.formattedResult); is throwing this error.

Regards.

**Comment by frozeman** [comment]
ok the tests seem to work, the only difference is, that the data is send as array, so could you change the last injected validation to:
```js
                    provider.injectValidation(function (payload) {
                        assert.equal(payload.jsonrpc, '2.0');
                        assert.equal(payload.method, 'eth_getFilterLogs');
                        assert.deepEqual(payload.params, [test.formattedResult]);
                    });
```

**Comment by bethnull** [comment]
yesterday I thought the same, the results are expected to be an array. However I tried it, and I got:

```
  1) web3.eth filter property test: 0:
     AssertionError: expected [ Array(2) ] to be a function
      at Context.<anonymous> (test/web3.eth.filter.js:84:28)
  ...
```
For the sake to make the tests work and against my way of doing things, I tried every possible combination of the arguments with the following results:

* assert.deepEqual(payload.params, test.formattedResult);
  * AssertionError: expected [ '0xf' ] to deeply equal '0xf'
* assert.deepEqual(payload.params, [test.formattedResult]);
  * AssertionError: expected [ Array(2) ] to be a function
* assert.deepEqual(payload.params[0], test.formattedResult);
  * AssertionError: expected [ Array(2) ] to be a function
* assert.deepEqual(payload.params[0], [test.formattedResult]);
  * AssertionError: expected '0xf' to deeply equal [ '0xf' ]

...

And I gave up and wrote to you ;)

**Comment by frozeman** [comment]
the line i send you is correct. the other error comes form line 84, which seems this line:

```js
assert.throws(filter.get());
```

This means, he passes the first validation in the validator, but now has a problem with simply calling `filter.get()`, which should throw an error.

I will pull your branch an take a look

**Comment by frozeman** [comment]
Ok i solved it now.

This is the correct file:
```js
var chai = require('chai');
var web3 = require('../index');
var assert = chai.assert;
var FakeHttpProvider = require('./helpers/FakeHttpProvider');

var method = 'filter';


var tests = [{
    args: [{
        fromBlock: 0,
        toBlock: 10,
        address: '0x47d33b27bb249a2dbab4c0612bf9caf4c1950855'
    }],
    formattedArgs: [{
        fromBlock: '0x0',
        toBlock: '0xa',
        address: '0x47d33b27bb249a2dbab4c0612bf9caf4c1950855',
        topics: []
    }],
    result: '0xf',
    formattedResult: '0xf',
    call: 'eth_newFilter'
},{
    args: [{
        fromBlock: 'latest',
        toBlock: 'latest',
        address: '0x47d33b27bb249a2dbab4c0612bf9caf4c1950855'
    }],
    formattedArgs: [{
        fromBlock: 'latest',
        toBlock: 'latest',
        address: '0x47d33b27bb249a2dbab4c0612bf9caf4c1950855',
        topics: []
    }],
    result: '0xf',
    formattedResult: '0xf',
    call: 'eth_newFilter'
},{
    args: ['latest'],
    formattedArgs: [],
    result: '0xf',
    formattedResult: '0xf',
    call: 'eth_newBlockFilter'
},{
    args: ['pending'],
    formattedArgs: [],
    result: '0xf',
    formattedResult: '0xf',
    call: 'eth_newPendingTransactionFilter'
}];

describe('web3.eth', function () {
    describe(method, function () {
        tests.forEach(function (test, index) {
            it('property test: ' + index, function () {
                
                // given
                var provider = new FakeHttpProvider();
                web3.setProvider(provider);
                provider.injectResult(test.result);
                provider.injectValidation(function (payload) {
                    assert.equal(payload.jsonrpc, '2.0');
                    assert.equal(payload.method, test.call);
                    assert.deepEqual(payload.params, test.formattedArgs);
                });

                // call
                var filter = web3.eth[method].apply(null, test.args);
                
                // test filter.get
                if(typeof test.args === 'object') {

                    var logs = [{data: '0xb'}, {data: '0x11'}];

                    provider.injectResult(logs);
                    provider.injectValidation(function (payload) {
                        assert.equal(payload.jsonrpc, '2.0');
                        assert.equal(payload.method, 'eth_getFilterLogs');
                        assert.deepEqual(payload.params, [test.formattedResult]);
                    });

                    // sync should throw an error
                    try {
                        assert.throws(filter.get());
                    } catch(e){
                        assert.instanceOf(e, Error);
                    }

                    // async should get the fake logs
                    filter.get(function(e, res){
                        assert.equal(logs, res);
                        done();
                    });
                }
            });
        });
    });
});
```

But please merge develop before into your branch as i had to make the async fake request helper really async.

**Comment by frozeman** [comment]
When you added the test and merged develop i will merge this commit. Thanks!

**Comment by bethnull** [comment]
Done, check it out and thanks for your invaluable help.

Regards, Pepe.

**Comment by frozeman** [comment]
there was a minor error in develop. could you please merge again and run the tests with `npm test`

**Comment by bethnull** [comment]
Done and all tests seems ok: 596 passing (4s)

**Comment by frozeman** [comment]
perfect! i will @debris take a look at it and we merge

**Comment by bethnull** [comment]
Cool. ;)

**Comment by debris** [comment]
looksgood! :+1:  I think it's a big step forward, although `filter.js` needs refactoring anyway :)

**Pull request merged**